components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/listener/JITProvisioningIdentityProviderMgtListener.java

```diff
@@ -0,0 +1,107 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.application.authentication.framework.handler.provisioning.listener;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.authentication.framework.handler.provisioning.ProvisionedUserDeleteThread;
+import org.wso2.carbon.identity.application.authentication.framework.util.FrameworkConstants;
+import org.wso2.carbon.identity.application.authentication.framework.util.FrameworkUtils;
+import org.wso2.carbon.identity.application.common.model.IdentityProvider;
+import org.wso2.carbon.identity.core.util.IdentityUtil;
+import org.wso2.carbon.idp.mgt.IdentityProviderManagementException;
+import org.wso2.carbon.idp.mgt.IdentityProviderManager;
+import org.wso2.carbon.idp.mgt.listener.AbstractIdentityProviderMgtListener;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Listener to perform JIT provisioning related Identity Provider management operations.
+ */
+public class JITProvisioningIdentityProviderMgtListener extends AbstractIdentityProviderMgtListener {
+
+    private static final Log log = LogFactory.getLog(JITProvisioningIdentityProviderMgtListener.class);
+    private static final Log diagnosticLog = LogFactory.getLog("diagnostics");
+    private static ExecutorService threadPool = Executors.newFixedThreadPool(2);
+
+    @Override
+    public boolean doPostDeleteIdPByResourceId(String resourceId, IdentityProvider identityProvider,
+                                               String tenantDomain) throws IdentityProviderManagementException {
+
+        if (!FrameworkUtils.isJITProvisionedEnhancedFeatureEnabled()) {
+            // JIT provisioning enhanced feature is not enabled. Hence returning.
+            return true;
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("The doPostDeleteIdPByResourceId executed for idp: %s of tenant domain: %s",
+                    resourceId, tenantDomain));
+        }
+
+        ProvisionedUserDeleteThread provisionedUserDeleteThread =
+                new ProvisionedUserDeleteThread(resourceId, tenantDomain);
+        threadPool.submit(provisionedUserDeleteThread);
+        return true;
+    }
+
+    @Override
+    public boolean doPreDeleteIdP(String idPName, String tenantDomain) throws IdentityProviderManagementException {
+
+        if (!FrameworkUtils.isJITProvisionedEnhancedFeatureEnabled()) {
+            // JIT provisioning enhanced feature is not enabled. Hence returning.
+            return true;
+        }
+
+        String idpId = IdentityProviderManager.getInstance().getIdPByName(idPName, tenantDomain,
+                true).getResourceId();
+        IdentityUtil.threadLocalProperties.get().put(FrameworkConstants.IDP_RESOURCE_ID, idpId);
```

Comment:
What happens if there is an issue in the idp deletition. Then the doPostDeleteIdP will not call and the thread-local may not clear

Reply:
Fixed with https://github.com/wso2/carbon-identity-framework/pull/3597/commits/0bb10a0d01ccec6f3ef0ba7ed7675e4fe360d177. Cleared the thread-local before retrieving it.